source/common/config/unified_mux/sotw_subscription_state.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include "envoy/grpc/status.h"
+#include "envoy/service/discovery/v3/discovery.pb.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/hash.h"
+#include "source/common/config/decoded_resource_impl.h"
+#include "source/common/config/unified_mux/subscription_state.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Config {
+namespace UnifiedMux {
+
+// Tracks the state of a "state-of-the-world" (i.e. not delta) xDS-over-gRPC protocol session.
+class SotwSubscriptionState
+    : public SubscriptionState<envoy::service::discovery::v3::DiscoveryResponse,
+                               envoy::service::discovery::v3::DiscoveryRequest> {
+public:
+  // Note that, outside of tests, we expect callbacks to always be a WatchMap.
+  SotwSubscriptionState(std::string type_url, UntypedConfigUpdateCallbacks& callbacks,
+                        std::chrono::milliseconds init_fetch_timeout, Event::Dispatcher& dispatcher,
+                        OpaqueResourceDecoder& resource_decoder);
+  ~SotwSubscriptionState() override;
+
+  // Update which resources we're interested in subscribing to.
+  void updateSubscriptionInterest(const absl::flat_hash_set<std::string>& cur_added,
+                                  const absl::flat_hash_set<std::string>& cur_removed) override;
+
+  // Whether there was a change in our subscription interest we have yet to inform the server of.
+  bool subscriptionUpdatePending() const override;
+
+  void markStreamFresh() override;
+
+  void ttlExpiryCallback(const std::vector<std::string>& expired) override;
+
+  SotwSubscriptionState(const SotwSubscriptionState&) = delete;
+  SotwSubscriptionState& operator=(const SotwSubscriptionState&) = delete;
+
+private:
+  std::unique_ptr<envoy::service::discovery::v3::DiscoveryRequest>
+  getNextRequestInternal() override;
+
+  void handleGoodResponse(const envoy::service::discovery::v3::DiscoveryResponse& message) override;
+  void setResourceTtl(const DecodedResourceImpl& decoded_resource);
+  bool isHeartbeatResource(const DecodedResource& resource, const std::string& version);
+
+  OpaqueResourceDecoder& resource_decoder_;
+
+  // The version_info carried by the last accepted DiscoveryResponse.
+  // Remains empty until one is accepted.
+  absl::optional<std::string> last_good_version_info_;
+  // The nonce carried by the last accepted DiscoveryResponse.
+  // Remains empty until one is accepted.
+  // Used when it's time to make a spontaneous (i.e. not primarily meant as an ACK) request.
+  absl::optional<std::string> last_good_nonce_;
```

Comment:
This is new stuff? I don't see these in the existing code. Any comments for reviewers on how to compare this to exisitng files?

Reply:
This is a substitute for calling `message.nonce()` directly; it makes passing `[Delta]DiscoveryResponse` unnecessary. It's used similarly to how message nonce is used in current GrpcMuxImpl::OnDiscoveryResponse (here: https://github.com/envoyproxy/envoy/blob/main/source/common/config/grpc_mux_impl.cc#L135). Please also see unifed mux implementation here: https://github.com/envoyproxy/envoy/pull/17049/commits/109ac32f8661ffed4dad2f607c6eec28b221435a#diff-fe56236f0649ba734162e4fa294465a6004c393242c0af89421cbd5fe279fe58R376

Reply:
Ack, makes sense.